### DISCUSSION TITLE
CppUTest versioning and conditional compilation

### DISCUSSION THREAD

**Discussion by mattagape** [cpputest/cpputest#1795] [Q&A]
Hi everyone

I've used CppUTest for a couple of years now. I'm mostly very pleased with it. It's been useful for regression testing, and finding memory leaks (although I do use Valgrind as well).

I've just been documenting the process I went through installing my build environment, etc. I am running Linux. I have not built CppUTest from the repo, but instead have installed the versions available with `apt`. Although I could do the former, I am putting together some "howto" instructions for my colleagues who aren't very familiar with some of the build processes.

Unfortunately, the versions installed using `apt` are different for the two flavours of Linux I am using. In both cases I believe I am using the latest versions available with `apt`. 

**On Linux Mint:**
```
$ apt-cache policy cpputest
cpputest:
  Installed: 3.8-7
  Candidate: 3.8-7
  Version table:
 *** 3.8-7 500
        500 http://archive.ubuntu.com/ubuntu focal/universe amd64 Packages
        500 http://archive.ubuntu.com/ubuntu focal/universe i386 Packages
        100 /var/lib/dpkg/status
```

**On Ubuntu:**
```
$  apt-cache policy cpputest
cpputest:
  Installed: 4.0-2
  Candidate: 4.0-2
  Version table:
 *** 4.0-2 500
        500 http://gb.archive.ubuntu.com/ubuntu jammy/universe amd64 Packages
        500 http://gb.archive.ubuntu.com/ubuntu jammy/universe i386 Packages
        100 /var/lib/dpkg/status
```
As you can see, the versions on the two versions of Linux are different. This matters because, for one set of tests, I switch off the memory leak checker, in the `teardown()` function:
```
void teardown()
{
   ...
   ...
    // This next line will fail if you are using an older version of CppUTest:
    MemoryLeakWarningPlugin::turnOnDefaultNotThreadSafeNewDeleteOverloads();
    // If so, use the next line instead:
    // MemoryLeakWarningPlugin::turnOnNewDeleteOverloads();
}
```
As indicated in the code comments, the version running on Linux Mint needs to use the function ` MemoryLeakWarningPlugin::turnOnNewDeleteOverloads()`, and the (newer) version on Ubuntu needs to use `MemoryLeakWarningPlugin::turnOnDefaultNotThreadSafeNewDeleteOverloads()`.

Rather than requiring the user to change the code as required, it would be nice if I could use a `#ifdef` directive to select the correct function based on the version of CppUTest running. Ideally the version in which this code change was made.

is it possible to do this? if so, where is the version number stored?

Thank you

Matthew

**Comment by Keith-H-Utron** [comment]
Looking at [getting-started](https://github.com/cpputest/cpputest#getting-started); it shows that you really do need to build this from the package source code.  Main reason is: you'll want MakefileWorker.mk to be included from your project's Makefile.  The best way for me to start up a new project is to follow James Grenning's step-by step instructions at [TDD Howto: Get your Legacy C into a Test Harness](https://wingman-sw.com/articles/tdd-legacy-c).  Central to that winning recipe is starting with the [Cpputest starter project](https://github.com/jwgrenning/cpputest-starter-project). Once you have the source package, you can read through the main package-generating files, configure.ac and Makefile.am. When used in the normal way with Autotools, output files are generated, including config.h. Here, you can find #define statements for PACKAGE_VERSION and VERSION. Include that $(CPPUTEST_HOME)/config.h and the rest is easy. However, I can't say how that works when building with Cmake.  Sorry it took a year for anyone to respond to you.  best regards, Keith
